.github/workflows/docker.yml

```diff
@@ -39,84 +54,181 @@ on:
-    if: github.event_name != 'schedule' || (github.event_name == 'schedule' && github.repository == 'jupyter/docker-stacks')
-    runs-on: ubuntu-latest
-    permissions:
-      contents: write
-
-    steps:
-      - name: Maximize build space 🛠
-        run: |
-          sudo rm -rf /usr/share/dotnet
-          sudo rm -rf /usr/local/lib/android
-          sudo rm -rf /opt/ghc
-
-      # Setup docker to build for multiple platforms, see:
-      # https://github.com/docker/build-push-action/tree/master#usage
-      # https://github.com/docker/build-push-action/blob/master/docs/advanced/multi-platform.md
-      - name: Set up QEMU (for docker buildx) 🐳
-        uses: docker/setup-qemu-action@8b122486cedac8393e77aa9734c3528886e4a1a8 # dependabot updates to latest release
-
-      - name: Set up Docker Buildx (for multi-arch builds) 🐳
-        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6 # dependabot updates to latest release
-
-      - name: Checkout Repo ⚡️
-        uses: actions/checkout@v3
-
-      - name: Set Up Python 🐍
-        uses: actions/setup-python@v3
-        with:
-          python-version: 3.x
-
-      - name: Install Dev Dependencies 📦
-        run: |
-          pip install --upgrade pip
-          pip install -r requirements-dev.txt
-
-      - name: Build Docker Images 🛠
-        run: make build-all-multi
-        env:
-          # Full logs for CI build
-          BUILDKIT_PROGRESS: plain
-
-      - name: Run tests ✅
-        run: make test-all
-
-      - name: Checkout Wiki Repo 📃
-        uses: actions/checkout@v3
-        with:
-          repository: ${{ github.repository }}.wiki
-          path: wiki/
-
-      - name: Create tags and manifest 🏷
-        run: make hook-all
-
-      - name: Push Wiki to GitHub 📤
-        if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' || github.event_name == 'schedule'
-        uses: stefanzweifel/git-auto-commit-action@49620cd3ed21ee620a48530e81dba0d139c9cb80 # dependabot updates to latest release
-        with:
-          commit_message: "Automated wiki publish for ${{ github.sha }}"
-          repository: wiki/
-
-      - name: Login to Docker Hub 🔐
-        if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' || github.event_name == 'schedule'
-        uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # dependabot updates to latest release
-        with:
-          username: ${{ secrets.DOCKERHUB_USERNAME }}
-          password: ${{ secrets.DOCKERHUB_TOKEN }}
-
-      - name: Push Images to Docker Hub 📤
-        if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' || github.event_name == 'schedule'
-        run: make push-all-multi
+  build-amd64-base:
+    uses: ./.github/workflows/docker-build-upload.yml
+    with:
+      parentImage: ""
+      image: base-notebook
+      platform: amd64
+      runsOn: ubuntu-latest
+
+  test-amd64-base:
```

Comment:
Actually in hindsight why is the build and test in separate jobs - so that you can kick - off building the child image as soon as the parent is built?

An alternative to avoid the save + load time accumulation might be to test right after build - then save it as artefact for use in other jobs. 
This might delay some downstream builds but might shave some time on the load task

Reply:
Yes, for example we now have a big dependency chain (which is probably the longest one):
base->minimal->scipy->pyspark->all-spark.

If we make testing inside the build step, then we will make this chain run for 5min49s longer (we will be running images).
At the same time the longest run of tests including downloading image is 5min44s.

So, there is practically no difference.

At the same time, if we add a general test, which runs X seconds, in my architecture it will add around X seconds to complete, but when we run inside the build step, it will add around 5X time.
So, current solution is better adapted for future test suite increase.

The times are based on this particular run (I hope I haven't made a mistake):
https://github.com/jupyter/docker-stacks/actions/runs/2321552950


Also, I think, having tests in the same step as a build is OK, and reduces the github graph size, so I might try it after we merge (or reject) this PR.